### PR TITLE
[RLlib] Move bandits into main agents folder; Make RecSim adapter more accessible;

### PR DESCRIPTION
**Pull request by sven1977**
- Move bandit algorithms into main rllib.agents folder (from rllib/contrib): LinTS and LinUCB
- Rename registry names for those algos from "contrib/Lin[TS|UCB]" into "BanditLin[TS|UCB]".
- Cleanup ResSim environment adapter and create 3 out-of-the-box RLlib-ready RecSim environments (interest evolution, long-term-satisfaction, and interest-exploration).
- Simple compilation tests for Bandits.

TODO (follow up PR):
- Run Benchmarks to use for hard-task nightly/weekly learning tests.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/bandits/bandit_torch_policy.py [hunk 1]
```diff
@@ -81,14 +78,14 @@ def make_model_and_action_dist(policy, obs_space, action_space, config):
 
     # TODO: Have a separate model catalogue for bandits
     if exploration_config:
-        if exploration_config["type"] == TS_PATH:
+        if exploration_config["type"] == "ThompsonSampling":
```
Comment (richardliaw) [inline rllib/agents/bandits/bandit_torch_policy.py:81]: maybe make this backwards compatible?

File: rllib/env/wrappers/tests/test_recsim_wrapper.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+    InterestEvolutionRecSimEnv
+from ray.rllib.env.wrappers.recsim import MultiDiscreteToDiscreteActionWrapper
 from ray.rllib.utils.error import UnsupportedSpaceException
 
 
 class TestRecSimWrapper(unittest.TestCase):
     def test_observation_space(self):
-        env = make_recsim_env(config={})
+        env = InterestEvolutionRecSimEnv()
         obs = env.reset()
```
Comment (avnishn) [inline rllib/env/wrappers/tests/test_recsim_wrapper.py:13]: we could also check if an observation that has been sampled from the observation space is contained, although im not sure how good a test that is. We could sampled a few thousand observations and check if they're contained in the obs space (its a relatively cheap operation)

**Review by avnishn — COMMENTED** [review]
LGTM, left a few comments, and asked clarifying questions, etc. Thanks for doing this, and all the hard work.

File: rllib/agents/bandits/tests/test_bandits.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+class TestBandits(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        ray.init()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        ray.shutdown()
+
+    def test_bandit_lin_ts_compilation(self):
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/bandits/tests/test_bandits.py:19]: +1
not necessarily this pr though.
like if we add tf impl for bandit, we can write unit tests to make sure those loss funcs are the same.

File: rllib/env/wrappers/recsim.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    def action(self, action: int) -> List[int]:
+        """Convert a Discrete action to a MultiDiscrete action"""
+        multi_action = [None] * len(self.action_space_dimensions)
+        for idx, n in enumerate(self.action_space_dimensions):
+            action, dim_action = divmod(action, n)
+            multi_action[idx] = dim_action
+        return multi_action
+
+
+def rllib_gym_wrapper(recsim_gym_env: gym.Env,
```
Comment (gjoliver) [inline rllib/env/wrappers/recsim.py:116]: recsys_gym_wrapper?

Comment (sven1977) [reply rllib/env/wrappers/recsim.py:116]: Oh no! Great catch! :D

File: rllib/agents/bandits/tests/test_bandits.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandits/tests/test_bandits.py:19]: Great idea @avnishn . Some of our algos do have these tests, e.g. ppo, dqn, pg, marwil. But not all. Yes, we should ideally have these for all algos. I agree that it helps to reduce bugs at a small cost!

File: rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py [hunk 5]
```diff
@@ -8,25 +8,28 @@
...
 
     print("Running training for %s time steps" % training_iterations)
 
     start_time = time.time()
     analysis = tune.run(
-        "contrib/LinUCB",
-        config=UCB_CONFIG,
+        "BanditLinUCB",
+        config=config,
         stop={"training_iteration": training_iterations},
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py:33]: Definitely need to clean up these scripts. These are the original ones from our contributors.

File: rllib/env/wrappers/recsim.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        An RLlib-ready gym.Env instance.
+    """
+    env = RecSimResetWrapper(recsim_gym_env)
+    env = RecSimObservationSpaceWrapper(env)
+    if convert_to_discrete_action_space:
+        env = MultiDiscreteToDiscreteActionWrapper(env)
+    return env
+
+
+def make_recsim_env(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/wrappers/recsim.py:152]: Yeah, I added/fixed the existing ones.

File: rllib/env/wrappers/recsim.py (hunk 4 above)
Comment (sven1977) [reply rllib/env/wrappers/recsim.py:116]: fixed

File: rllib/env/wrappers/tests/test_recsim_wrapper.py (hunk 2 above)
Comment (sven1977) [reply rllib/env/wrappers/tests/test_recsim_wrapper.py:13]: I think we are doing this right below. Only twice, though after the `env.reset()` and the `step()`.

Added some more checks for the action space test.

Not sure I fully understand what you are aiming at, though. But please feel free to suggest more checks here.

File: rllib/agents/bandits/bandit_torch_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/bandits/bandit_torch_policy.py]: Not sure this would work: The exploration type is now a single "ThompsonSampling" string as we moved the `ThompsonSampling` class into our built-in RLlib `utils/exploration` package.

The two supported exploration types are pretty much hard-coded into these two bandit Trainers (UCB and TS).
Also, the old `TS_PATH` ( -> ThompsonSampling path) does no longer exist as it was inside the contrib folder.

File: rllib/agents/bandits/bandit_torch_policy.py [hunk 7]
```diff
@@ -81,14 +78,14 @@ def make_model_and_action_dist(policy, obs_space, action_space, config):
...
             if isinstance(original_space, spaces.Dict):
                 assert "item" in original_space.spaces, \
                     "Cannot find 'item' key in observation space"
                 model_cls = ParametricLinearModelThompsonSampling
             else:
                 model_cls = DiscreteLinearModelThompsonSampling
-        elif exploration_config["type"] == UCB_PATH:
+        elif exploration_config["type"] == "UpperConfidenceBound":
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandits/bandit_torch_policy.py]: Please see my comment above.

File: rllib/agents/bandits/bandit_torch_policy.py (hunk 1 above)
Comment (richardliaw) [reply rllib/agents/bandits/bandit_torch_policy.py]: OK, this may not be the right level to do backward compat messaging. Can you make sure that users code from previous ray versions will automatically raise an actionable error message to help them migrate to new changes?

Comment (sven1977) [reply rllib/agents/bandits/bandit_torch_policy.py]: This is done through the added "location" check. The new `test_bandits` test case makes sure this works and users will get a deprecation message when they do stuff like:

```
tune.run("contrib/LinTS")

# OR

from ray.rllib.contrib.bandits.agents.lin_ts import LinTSTrainer
```

File: rllib/agents/bandit/tests/test_bandits.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        num_iterations = 5
+
+        for _ in framework_iterator(config, frameworks="torch"):
+            trainer = bandit.BanditLinTSTrainer(config=config)
+            results = None
+            for i in range(num_iterations):
+                results = trainer.train()
+                check_train_results(results)
```
Comment (avnishn) [inline rllib/agents/bandit/tests/test_bandits.py:36]: Can we seed this test.

**Review by avnishn — APPROVED** [review]
Had some comments about seeding  learning tests but otherwise lgtm.

File: rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py]: Can we do this in a follow up PR?
Just moved these here, these are not new scripts. Let's go through them in another PR and make sure the Bandits learn properly.

File: rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py [hunk 9]
```diff
@@ -8,25 +8,28 @@
 import pandas as pd
 
 from ray import tune
-from ray.rllib.contrib.bandits.agents.lin_ucb import UCB_CONFIG
-from ray.rllib.contrib.bandits.envs import ParametricItemRecoEnv
+from ray.rllib.examples.env.bandit_envs_recommender_system import \
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/bandits/tests/tune_LinUCB_train_recommendation.py]: Sorry, missed this:
Wait, it's there, no?

```
py_test(
    name = "test_bandits",
    tags = ["team:ml", "trainers_dir"],
    size = "small",
    srcs = ["agents/bandit/tests/test_bandits.py"],
)

py_test(
    name = "bandits/tests/lin_ts_train_wheel_env",
    main = "bandits/tests/LinTS_train_wheel_env.py",
    tags = ["team:ml", "trainers_dir"],
    size = "small",
    srcs = ["agents/bandit/tests/LinTS_train_wheel_env.py"],
)

py_test(
    name = "bandits/tests/tune_lin_ts_train_wheel_env",
    main = "bandits/tests/tune_LinTS_train_wheel_env.py",
    tags = ["team:ml", "trainers_dir"],
    size = "small",
    srcs = ["agents/bandit/tests/tune_LinTS_train_wheel_env.py"],
)

py_test(
    name = "bandits/tests/tune_lin_ucb_train_recommendation",
    main = "bandits/tests/tune_LinUCB_train_recommendation.py",
    tags = ["team:ml", "trainers_dir"],
    size = "small",
    srcs = ["agents/bandit/tests/tune_LinUCB_train_recommendation.py"],
)
```

**Review by gjoliver — APPROVED** [review]
lgtm
